### PR TITLE
Experiment: Move unselected block accessibility handling to block-holder

### PR DESCRIPTION
**Pull request by Tug**
Alternative to https://github.com/wordpress-mobile/gutenberg-mobile/pull/936
Gutenberg PR https://github.com/WordPress/gutenberg/pull/15225

> This PR moves accessibility handling of unselected blocks to the BlockHolder

It updates e2e tests as well which were broken with this updates (we rely on the accessibility label to find blocks).

### Testing Instructions

- Try navigating blocks on Android with Talkback on
- Try navigating blocks on iOS with VoiceOver on

For instance:

- An heading block should have all the following information spoken:
"Heading Block. Row X. <content>. Button"

- A paragraph block should have all the following information spoken:
"Paragraph Block. Row X. <content>. Button"

- An empty image block should have all the following information spoken:
"Image Block. Row X. Empty. Button"

- A non-empty image block with a caption should have all the following information spoken:
"Image Block. Row X. <content of the caption>. Button"

### CODE REVIEW

File: src/block-management/block-holder.js [hunk 1]
```diff
@@ -197,6 +198,28 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
 		);
 	}
 
+	getAccessibilityLabel() {
+		const { title, attributes, name } = this.props;
+
+		if ( name === getUnregisteredTypeHandlerName() ) { // is the block unrecognized?
+			return title; //already localized
+		}
```
[the remark this answers is not in the document]
Comment (etoledom) [reply src/block-management/block-holder.js]: I don't mean to add accessibility labels to `missing` block on this PR, but just saying that it might not need an special handling.

File: src/block-management/block-holder.js [hunk 2]
```diff
@@ -206,8 +229,9 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
 			// accessible prop needs to be false to access children
 			// https://facebook.github.io/react-native/docs/accessibility#accessible-ios-android
 			<TouchableWithoutFeedback
-				accessible={ false }
-				accessibilityLabel="block-container"
+				accessible={ ! isSelected }
```
Comment (JavonDavis) [inline src/block-management/block-holder.js]: I like this, it makes more sense to only make it false when it's selected and we'll only interact with the inner elements when it's selected anyway

**Comment by JavonDavis** [comment]
>  I wonder how this could impact UITests

It shouldn't and didn't 🙂Since the `TouchableWithoutFeedback` is set to `false` when the block is selected, the tests should continue as they are.

**Comment by etoledom** [comment]
As a note: This PR fixes #14, #910 and (probably) #950

File: src/block-management/block-holder.js (hunk 2 above)
Comment (Tug) [reply src/block-management/block-holder.js]: Yes, it doesn't seem to work on android though...

File: src/block-management/block-holder.js [hunk 3]
```diff
@@ -197,6 +198,28 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
...
+		const blockName = sprintf(
+			/* translators: accessibility text. %s: block name. */
+			_x( '%1$s block' ),
+			title, //already localized
+		);
+
+		const blockType = getBlockType( name );
+		if ( blockType.getAccessibilityLabel ) {
+			const blockAccessibilityLabel = blockType.getAccessibilityLabel( attributes ) || '';
+			return blockName + ( blockAccessibilityLabel ? '. ' + blockAccessibilityLabel : '' );
```
[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-holder.js]: I am aiming towards to the current approach because it brings consistency. With this, the description for all blocks starts with 'X block. '

on the other hand I think the '. ' and concatenations needs a revisit as we discussed earlier for localization issues. would it be better to do this as follows?

```
             sprintf(
			/* translators: accessibility text. 1: block name. 2: block content information. */
			_x( '%1$s block. %2$s', 'Accessibility text for a block' ),
			blockName,
			blockAccessibilityLabel
		);
```

File: src/block-management/block-holder.js [hunk 4]
```diff
@@ -163,6 +165,28 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
...
+	getAccessibilityLabel() {
+		const { title, attributes, name } = this.props;
+
+		if ( name === getUnregisteredTypeHandlerName() ) { // is the block unrecognized?
+			return title; //already localized
+		}
+
+		const blockName = sprintf(
+			/* translators: accessibility text. %s: block name. */
+			_x( '%1$s block' ),
```
Comment (etoledom) [inline src/block-management/block-holder.js]: Should this be `__` instead of `_x`?
And since there's just one variable, `%d` should be enough?

**Comment by Tug** [comment]
> I think the getAccessibilityExtra call isn't working as expected.

File: src/block-management/block-holder.js [hunk 5]
```diff
@@ -174,31 +167,56 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
...
 			<TouchableWithoutFeedback
-				accessible={ false }
 				onPress={ this.onFocus }
 			>
-
 				<View style={ [ styles.blockHolder, borderStyle, { borderColor } ] }>
 					{ this.props.showTitle && this.renderBlockTitle() }
 					<View
-						accessibile={ true }
+						accessible={ ! isSelected }
```
[the remark this answers is not in the document]
Comment (Tug) [reply src/block-management/block-holder.js]: Yeah I wanted to basically group all the accessibility props together and View already had the `accessibilityLabel` prop.
The idea is that both `TouchableWithoutFeedback` and the `View` are containers of our block so both have the potential to be accessible elements. I did some test and I'm not sure it changes things much where we put it.

File: src/block-management/block-holder.js [hunk 6]
```diff
@@ -174,31 +167,56 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
...
 						accessibilityLabel={ accessibilityLabel }
+						accessibilityRole={ 'button' }
```
[the remark this answers is not in the document]
Comment (Tug) [reply src/block-management/block-holder.js]: This is only meant for the screenreader. Adding `accessibilityRole={ 'button' }` will translate to the screenreader saying "button" at after reading the `accessibilityLabel`. Meaning the user can interact on it like a normal button (ouble tap to select the block in this case)

**Comment by Tug** [comment]
> I think the getAccessibilityExtra call isn't working as expected.

This would be annoying because it's the main change of this PR. I've checked again and it seems to be working. Could you expand on what's not working here?

Anyway it's green now thanks to your commit https://github.com/wordpress-mobile/gutenberg-mobile/pull/937/commits/a55b735c837a180d5f49e73b21a4e46df078bd0b Thanks a lot @JavonDavis 🙇 

@etoledom @pinarol Could you give it a try now?

File: src/block-management/block-holder.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (JavonDavis) [reply src/block-management/block-holder.js]: > I did some test and I'm not sure it changes things much where we put it.

Oh, I played around with it too and could've sworn it was affecting some things for me but I was playing with a few things and I don't remember for certain, let me try again and see if I can reproduce again here

**Comment by JavonDavis** [comment]
> This would be annoying because it's the main change of this PR. I've checked again and it seems to be working. Could you expand on what's not working here?

Sorry for taking a little while to clarify what I meant here... When I was testing this locally and running  the e2e tests the code in the `if ( getAccessibilityLabelExtra ) {` statement was being executed and from my understanding of the comment above that line it shouldn't(?) and that's what caused the need for the new XPath using `contains`? removing that like I did in #1041 I was able to go back to the `=` instead, I'm going to test this again to be sure...

File: src/block-management/block-holder.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-holder.js]: Hey @Tug 👋 I tested this and currently it is not saying "button" after reading `accessibilityLabel`. But when I move `accessibilityRole={ 'button' }` to `TouchableWithoutFeedback` level it does say it:

```
			<TouchableWithoutFeedback
				onPress={ this.onFocus }
				accessible={ ! isSelected }
				accessibilityRole={ 'button' }
			>
```

So we might want to update this like that.

[the remark this answers is not in the document]
Comment (Tug) [reply src/block-management/block-holder.js]: Indeed it does that on iOS when the block is unselected. I was able to confirm that moving `accessibilityRole="button"` to `TouchableWithoutFeedback`  fixes it on both platforms. However, I also I noticed that it says "button" first before reading the view's accessibility label on Android so I moved everything to `TouchableWithoutFeedback` and now it works as expected!

[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-holder.js]: It looks like CI tests are failing this time. I have an idea, we can just set an `accessibilityHint` rather than an accessibilityRole and keep the accessibilityLabel where it was. 

```
			<TouchableWithoutFeedback
				onPress={ this.onFocus }
				accessible={ ! isSelected }
				accessibilityHint={  __('Double tap to focus')  }
			>
				<View style={ [ styles.blockHolder, borderStyle, { borderColor } ] }>
					{ this.props.showTitle && this.renderBlockTitle() }
					<View
						accessibilityLabel={ accessibilityLabel }
```

[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-holder.js]: Another option would be ofc updating the CI tests according to the current settings

**Comment by JavonDavis** [comment]
>  I'm going to test this again to be sure...

@Tug My bad, it seems the issue I was seeing was related to something else, `getAccessibilityLabelExtra` works as expected! 👍

File: src/block-management/block-holder.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tug) [reply src/block-management/block-holder.js]: Tried to debug the e2e tests but as I was wasting too much time on this I ended up reverting the change in [this commit](https://github.com/wordpress-mobile/gutenberg-mobile/pull/937/commits/ab3f0a35ffb87d75f9e04c97a2e936f77247f9f4).
I guess it's not too critical to have "button" mentioned before the accessibility label on android. We have many improvements to do on accessibility anyway.
@pinarol could you give it a final check please :)?

[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-holder.js]: At least the issue is resolved on iOS, now I hear: "Paragraph block. Row n. Text. Button" 👍 Thanks
